### ISSUE TITLE
Allow notification plugins to notify about the pipeline status

### ISSUE THREAD

**Issue by xoxys** [woodpecker-ci/woodpecker#4337]
### Clear and concise description of the problem

The env var `CI_PIPELINE_STATUS` war broken already and has always reported `success` but was now removed completely in https://github.com/woodpecker-ci/woodpecker/pull/3846

Right now, I don't see any way to use a notification plugin in pipelines to send build status notifications to an external system.

### Suggested solution

Add back `CI_PIPELINE_STATUS` and populate the correct status. We have the correct pipeline status in the UI/API already, why can't we just populate it to the env?

### Validations

- [X] Checked that the feature isn't part of the `next` version already [https://woodpecker-ci.org/faq#which-version-of-woodpecker-should-i-use]
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't already an [issue](https://github.com/woodpecker-ci/woodpecker/issues) that request the same feature to avoid creating a duplicate.

**Comment by zc-devs** [comment]
> CI_PIPELINE_STATUS and populate the correct status

Will always be `running` :)

**Comment by xoxys** [comment]
Well coming from Drone I can at least tell you there was a working and stable implementation.

**Comment by zc-devs** [comment]
> there was a working and stable implementation

Sure. I didn't say that it is impossible. Do you know how it's implemented there?

Could you show an example pipeline? Does watched step have to define `failure: ignore`?

If we are using `failure: ignore` and step within the same pipeline, then one of implementation could be var like `CI_STEPS_STATUS` with value like `clone:success,test:success,build:fail`. Or separate variables `CI_STEP_CLONE_STATUS:success` and so on.

Another solution could be webhooks to external service.

**Comment by xoxys** [comment]
> Sure. I didn't say that it is impossible. Do you know how it's implemented there?

Drone uses some kind of hooks https://github.com/harness/harness/blob/drone/operator/runner/runner.go#L380-L402 which looks kinda similar to what we had in [place ](https://github.com/woodpecker-ci/woodpecker/pull/3846/files#diff-44974d0d24afd379d977e3ef7cd2a73fb5449937b64e82b6578555c90fc35be8) already. That's why I'm wondering why it was broken in Woodpecker...

> Could you show an example pipeline? Does watched step have to define failure: ignore?

https://github.com/thegeeklab/wp-matrix/blob/main/.woodpecker/notify.yml

> Yet another one: use special step type likeservice, but notification. It would always run after pipeline finishes and yes can contain correct CI_PIPELINE_STATUS.

Sure could be done this way also.

**Comment by zc-devs** [comment]
Thank you for an example. Wasn't aware of `runs_on`. It's like my `special step type` :)
Agree that it was/is a bug and should be fixed.

**Comment by xoxys** [comment]
The agent is pulling the next workflow to execute from the queue but I cant see any way to access the parent pipeline information from the agent. Do I miss something? Is there any existing way to lookup the pipeline e.g. with the `workflow.PipelineID` from the agent?

**Comment by zc-devs** [comment]
You can probably look it up via REST and even make custom GRPC method, but why It's the server responsibility to set (almost) all [vars](https://github.com/woodpecker-ci/woodpecker/blob/main/pipeline/frontend/metadata/environment.go). And this map sends to the agent.

**Comment by xoxys** [comment]
This map is only set once during the pipeline creation and can only contain static values that dont change during workflow executions. For the pipeline state it must be looked up before a workflow is pulled from the queue and executed.

**Comment by zc-devs** [comment]
Notifications have to be executed after all ... ~what actually (steps in workflow or workflows)? Do you want notification per workflow or per pipeline?~

1. ~Per pipeline.~ I think you can now define the separate notifications workflow as `depends_on` other in a way it executes the last (after workflows you want notifications about). So at the time the notifications workflow is ready to execute, the Server will know the statuses of all other executed workflows. The server sets up env `CI_WORKFLOW_STATUSES` (or separate ones) and sends the notifications workflow into queue. Agent picks it up and all works.
2. It actually works for both workflows statuses and pipeline. So you have statuses of each workflow. And if all workflows are success, then the pipeline is also success -> you can set `PIPELINE_STATUS` on the server a bit in advance.

**Comment by xoxys** [comment]
Maybe I miss something, but I don't see how your approach should work:

1) Pipeline is triggered by event
2) Pipeline config will be parsed, and all workflows will be compiled and added to the queue. At this point all static env vars from the location you have linked are already added to the workflow metadata before they are added to the queue.
3) Workflows are executed, one or more failed but the notification workflow which is already in the queue at this point doesnt know about the changed pipeline state (how should it?)

**Comment by zc-devs** [comment]
> all workflows will be compiled and added to the queue
> the notification workflow which is already in the queue

All except notifications. Compile ... maybe, but not put in queue. That was the point in my approach. Agree, that now it might not work, but it could be some mark in syntax to define such workflows.

And still, I wonder
1. Does the examples work at all?
2. How do they work then? Because before we (server?) run a step with `when.status:failure` we need `failure` of some "entity".

**Comment by 6543** [comment]
the current pipeline status is just technically **undefined** till it has finished.

and `CI_PREV_PIPELINE_STATUS` is still here ... so i dont get what the problem is ...

also taling abaut notification plugin you might wana have a look into CI_PREV_STEP_STATUS

**Comment by xoxys** [comment]
Goal: At the end of a **pipeline** I want to have a step that sends the ** current pipeline** status (success or failed) to an external system.  Is this possible right now?

**Comment by zc-devs** [comment]
It should have been possible with
```yaml
# build.yaml
skip_clone: true
steps:
  build:
    image: alpine
    commands:
      - echo 'Building the app'
      - exit 1 # 0 - success, 1 - fail
```
```yaml
# notifications.yaml
skip_clone: true
depends_on: [build]
runs_on: [success, failure]
steps:
  fail-notification:
    when:
      - status: [failure]
    image: alpine
    commands:
      - echo 'Build failed'
  success-notification:
    when:
      - status: [success]
    image: alpine
    commands:
      - echo 'Build succeed'
```
But it doesn't work
![Screenshot 2024-11-11 1](https://github.com/user-attachments/assets/77cad464-54ff-4aef-996a-a18d6ff5ad33)

**Comment by xoxys** [comment]
This is a separate bug IMO. But to make it a bit more clear, even if this bug doesn't exist, this way is a more hack and not proper solution for notifications.

**Comment by ambroisie** [comment]
When did `CI_PIPELINE_STATUS` stop working? It's been working reliably for me since before 2.0 and still is on 2.7.3 (I use it as part of a [notifier script](https://github.com/ambroisie/matrix-notifier), running both on failure and success).

**Comment by xoxys** [comment]
See the linked PR in the issue description.

**Comment by ambroisie** [comment]
I first found the PR thanks to the release notes of the 3.0 release candidate, and _then_ this issue: I didn't see any bisection/mention of when the variable was broken.

**Comment by xoxys** [comment]
Thats the purpose of this issue. Notifications are no longer supported in WP 3.x

**Comment by anbraten** [comment]
Should be pretty easy to bring back by adding the state around here: https://github.com/woodpecker-ci/woodpecker/blob/e053cd44d96baad454d64d8281bd9437e7e793d0/agent/tracer.go#L73

**Comment by xoxys** [comment]
IIRC this was the place where it was removed and because it was set to "successful" even if the pipeline failed.

**Comment by anbraten** [comment]
Will need to test, but I am pretty sure parts of it, at least the slack plugin sending a message on pipeline errors was working before.

**Comment by xoxys** [comment]
The slack plugin works in the same way as all others and dont rely on the env var directly. You can still sent notifications but getting the pipeline status into the message that is sent by the plugin doesnt work anymore.

As a hack two steps can be used, one with 

```
    when:
      status: [success]
```

anf one with 

```
    when:
      status: [failure]
```

You would have to use a hardcoded status in the message template (if supported by the plugin) to get the correct pipeline status into your messages.

As I said, the issue with the removed env var was that the value was always "success" which makes notificatios that are sent on success and failure events useless.

**Comment by xoxys** [comment]
Not sure if asking every few weeks might help. Right now nobody has start working on it.

**Comment by kapawit** [comment]
im also having this issue, please fix it

**Comment by 6543** [comment]
with the latest refactoring, you can bring back `CI_PIPELINE_STATUS` savely :)

**Comment by 6543** [comment]
yes was easy: https://github.com/woodpecker-ci/woodpecker/pull/6516 also finished refactoring the step env set there...

**Comment by 6543** [comment]
> ...  What I want to achieve here is push the topic forward cause it looks like it was stale for some time now.

sorry the code just had to be refactored to make sense at all 😅
